apps/btc_app/get_public_key.c

```diff
@@ -0,0 +1,292 @@
+  uint8_t address_version = 0;
+  uint32_t purpose_index = path[0];
+  uint32_t coin_index = path[1];
+  char addr[50] = "";
+  size_t address_length = 0;
+
+  if (!derive_hdnode_from_path(
+          path, path_length, SECP256K1_NAME, seed, &node)) {
+    // send unknown error; unknown failure reason
+    btc_send_error(ERROR_COMMON_ERROR_UNKNOWN_ERROR_TAG, 1);
+    memzero(&node, sizeof(HDNode));
+    return address_length;
+  }
+
+  switch (purpose_index) {
+    case NATIVE_SEGWIT:
+      // ignoring the return status and handling by size of address
+      get_segwit_address(
+          node.public_key, sizeof(node.public_key), coin_index, addr);
+      break;
+    case NON_SEGWIT:
+      get_version(purpose_index, coin_index, &address_version, NULL);
+      hdnode_get_address(&node, address_version, addr, 35);
+      break;
+    // TODO: add support for taproot and segwit
+    default:
+      break;
+  }
+
+  address_length = strnlen(addr, sizeof(addr));
+  if (0 >= address_length) {
+    // address encoding failed
+    btc_send_error(ERROR_COMMON_ERROR_UNKNOWN_ERROR_TAG, 1);
+  }
+  if (NULL != public_key) {
+    ecdsa_uncompress_pubkey(
+        get_curve_by_name(SECP256K1_NAME)->params, node.public_key, public_key);
+  }
+  if (NULL != address) {
+    memcpy(address, addr, address_length);
+  }
+  memzero(&node, sizeof(HDNode));
+  return address_length;
+}
+
+static void send_public_key(const uint8_t *public_key) {
+  btc_result_t response = init_btc_result(BTC_RESULT_GET_PUBLIC_KEY_TAG);
+  response.get_public_key.which_response =
+      BTC_GET_PUBLIC_KEY_RESPONSE_RESULT_TAG;
+  ecdsa_uncompress_pubkey(get_curve_by_name(SECP256K1_NAME)->params,
+                          public_key,
+                          response.get_public_key.result.public_key);
+  btc_send_result(&response);
+}
+
+/*****************************************************************************
+ * GLOBAL VARIABLES
+ *****************************************************************************/
+
+/*****************************************************************************
+ * STATIC FUNCTIONS
+ *****************************************************************************/
+
+/*****************************************************************************
+ * GLOBAL FUNCTIONS
+ *****************************************************************************/
+
+void btc_get_public_key(btc_query_t *query) {
+  const char *wallet_name = "";
+  char string[100] = "";
+  uint8_t wallet_index = 0;
+  uint8_t public_key[65] = {0};
+  btc_get_public_key_intiate_request_t *init_req =
+      &query->get_public_key.initiate;
+
+  if (!check_which_request(query, BTC_GET_PUBLIC_KEY_REQUEST_INITIATE_TAG) ||
+      !validate_request_data(&query->get_public_key)) {
+    return;
+  }
+
+  // this will always succeed; can skip checking return value
+  get_first_matching_index_by_id(init_req->wallet_id, &wallet_index);
+  wallet_name = (const char *)get_wallet_name(wallet_index);
```

Comment:
Add a todo to replace this handling with a single function that returns name from wallet id

Reply:
added